scenario-compiler/src/main/scala/pl/touk/nussknacker/engine/extension/ExtensionMethods.scala

```diff
@@ -1,44 +1,54 @@
 package pl.touk.nussknacker.engine.extension
 
-import pl.touk.nussknacker.engine.definition.clazz.ClassDefinitionSet
+import pl.touk.nussknacker.engine.definition.clazz.{ClassDefinitionSet, MethodDefinition}
+import pl.touk.nussknacker.engine.extension.ExtensionMethods.extensions
 
 import java.lang.reflect.Method
 
+class ExtensionMethodsInvoker(classLoader: ClassLoader, classDefinitionSet: ClassDefinitionSet) {
+  private val extensionsByClass = extensions.map(e => e.clazz -> e).toMap[Class[_], Extension]
+
+  def invoke(target: Object, arguments: Array[Object]): PartialFunction[Method, Any] = {
+    case method if extensionsByClass.contains(method.getDeclaringClass) =>
+      extensionsByClass
+        .get(method.getDeclaringClass)
+        .map(_.implFactory.create(target, classLoader, classDefinitionSet))
+        .map(impl => method.invoke(impl, arguments: _*))
+        .getOrElse {
+          throw new IllegalArgumentException(s"Extension method: ${method.getName} is not implemented")
+        }
+  }
+
+}
+
 object ExtensionMethods {
 
-  private val declarationsWithImplementations = Map[Class[_], ExtensionMethodsImplFactory](
-    classOf[Cast] -> CastImpl,
+  private[extension] val extensions = List(
+    Extension(classOf[Cast], Cast, Cast)
   )
 
-  private val registry: Set[Class[_]] = declarationsWithImplementations.keySet
-
   def enrichWithExtensionMethods(set: ClassDefinitionSet): ClassDefinitionSet = {
-    val castMethodDefinitions = CastMethodDefinitions(set)
     new ClassDefinitionSet(
       set.classDefinitionsMap.map { case (clazz, definition) =>
-        clazz -> definition.copy(methods = definition.methods ++ castMethodDefinitions.extractDefinitions(clazz))
+        clazz -> definition.copy(
+          methods = definition.methods ++ extensions.flatMap(_.definitionsExtractor.extractDefinitions(clazz, set))
+        )
       }.toMap // .toMap is needed by scala 2.12
     )
   }
 
-  def invoke(
-      method: Method,
-      target: Any,
-      arguments: Array[Object],
-      classLoader: ClassLoader
-  ): PartialFunction[Class[_], Any] = {
-    case clazz if registry.contains(clazz) =>
-      declarationsWithImplementations
-        .get(method.getDeclaringClass)
-        .map(_.create(target, classLoader))
-        .map(impl => method.invoke(impl, arguments: _*))
-        .getOrElse {
-          throw new IllegalArgumentException(s"Extension method: ${method.getName} is not implemented")
-        }
-  }
-
 }
 
 trait ExtensionMethodsImplFactory {
-  def create(target: Any, classLoader: ClassLoader): Any
+  def create(target: Any, classLoader: ClassLoader, set: ClassDefinitionSet): Any
 }
+
+trait ExtensionMethodsDefinitionsExtractor {
+  def extractDefinitions(clazz: Class[_], set: ClassDefinitionSet): Map[String, List[MethodDefinition]]
+}
+
+case class Extension(
```

Comment:
it's not a data class, so IMO it's better to remove `case`

Reply:
Done